content/en/blog/_posts/2025-02-14-cloud-controller-manager-chicken-egg-problem.md

```diff
@@ -0,0 +1,124 @@
+---
+layout: blog
+title: "The Cloud Controller Manager Chicken and Egg Problem"
+date: 2025-02-14
+slug: cloud-controller-manager-chicken-egg-problem
+author: >
+  Antonio Ojea,
+  Michael McCune
+draft: true
+---
+
+Kubernetes 1.31 [completed the largest migration in Kubernetes history](https://kubernetes.io/blog/2024/05/20/completing-cloud-provider-migration/), removing the in-tree cloud provider.  While the component migration is now done, this leaves some additional complexity for users and installer projects (for example, kOps or Cluster API) .  We will go over those additional steps and failure points and make recommendations for cluster owners.  
+This migration was complex and some logic had to be extracted from the core components, building four new subsystems.
+
+1. **Cloud controller manager** ([KEP-2392](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2392-cloud-controller-manager/README.md))  
+2. **API server network proxy** ([KEP-1281](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1281-network-proxy))  
+3. **kubelet credential provider plugins** ([KEP-2133](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2133-kubelet-credential-providers))  
+4. **Storage migration to use [CSI](https://github.com/container-storage-interface/spec?tab=readme-ov-file#container-storage-interface-csi-specification-)** ([KEP-625](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/625-csi-migration/README.md))
+
+The [cloud controller manager is part of the control plane](https://kubernetes.io/docs/concepts/architecture/cloud-controller/). It is a critical component that replaces some functionality that existed previously in the kube-controller-manager and the kubelet.
+
+![Components of Kubernetes](https://kubernetes.io/images/docs/components-of-kubernetes.svg)
+
+One of the most critical functionalities of the cloud controller manager is the node controller, which is responsible for the initialization of the nodes.
+
+As we can see in the following diagram, when the **kubelet** starts, it registers the `Node` object with the apiserver, Tainting the node so it can be processed first by the cloud-controller-manager. The initial `Node` is missing the cloud-provider specific information, like the Node Addresses and the Labels with the cloud provider specific information like the Node, Region and Instance type information.
```

Comment:
We avoid “we” unless you specifically mean “we,   Antonio Ojea & Michael McCune”.